### PR TITLE
Move ICompilationException to Diagnostics

### LINKED ISSUE #220: Move ICompilationException to Diagnostics.Abstractions
We use `ICompilationException` as a contract between diagnostics and other middlewares (primarily Mvc) that perform runtime compilation to show a pretty printed compilation failure. Currently the contract for this exception is located in `Microsoft.Extensions.CompilationAbstractions`. This would allow removing the CompilationAbstractions dependency.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #220

### CODE REVIEW

File: src/Microsoft.AspNet.Diagnostics.Abstractions/DiagnosticMessage.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNet.Diagnostics
+{
+    /// <summary>
+    /// A single diagnostic message.
+    /// </summary>
+    public class DiagnosticMessage
+    {
+        public DiagnosticMessage(
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Diagnostics.Abstractions/DiagnosticMessage.cs]: I removed some of the properties from this - `DiagnosticMessageSeverity`, `ErrorCode` and `object source`. Not sure if there's value in keeping the first - the latter two were used in tooling scenarios.

**Comment by pranavkm** [comment]
cc @davidfowl

File: src/Microsoft.AspNet.Diagnostics/project.json [hunk 2]
```diff
@@ -14,9 +14,9 @@
     "Microsoft.AspNet.FileProviders.Physical": "1.0.0-*",
     "Microsoft.AspNet.Http.Extensions": "1.0.0-*",
     "Microsoft.AspNet.WebUtilities": "1.0.0-*",
-    "Microsoft.Extensions.CompilationAbstractions": "1.0.0-*",
     "Microsoft.Extensions.Logging.Abstractions": "1.0.0-*",
     "Microsoft.Extensions.OptionsModel": "1.0.0-*",
+    "Microsoft.Extensions.PlatformAbstractions": "1.0.0-*",
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Diagnostics/project.json]: Don't use any types from here. Copy those over too.

Comment (pranavkm) [inline src/Microsoft.AspNet.Diagnostics/project.json]: There're types in here which use `ILibraryManager` - https://github.com/aspnet/Diagnostics/blob/dev/src/Microsoft.AspNet.Diagnostics/RuntimeInfo/RuntimeInfoMiddleware.cs#L81. Suggestions on what to do with that?

Comment (davidfowl) [inline src/Microsoft.AspNet.Diagnostics/project.json]: Ah, that's ok for now.

File: src/Microsoft.AspNet.Diagnostics/DeveloperExceptionPage/DeveloperExceptionPageMiddleware.cs [hunk 3]
```diff
@@ -4,19 +4,16 @@
...
 using System.Linq;
 using System.Threading.Tasks;
-using Microsoft.AspNet.Builder;
 using Microsoft.AspNet.Diagnostics.Views;
 using Microsoft.AspNet.FileProviders;
 using Microsoft.AspNet.Http;
-using Microsoft.Extensions.CompilationAbstractions;
-using Microsoft.Extensions.PlatformAbstractions;
 using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.PlatformAbstractions;
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Diagnostics/DeveloperExceptionPage/DeveloperExceptionPageMiddleware.cs]: What is used from this namespace?

Comment (pranavkm) [inline src/Microsoft.AspNet.Diagnostics/DeveloperExceptionPage/DeveloperExceptionPageMiddleware.cs]: `IAppEnv` for the base path. We can replace it with code

Comment (davidfowl) [inline src/Microsoft.AspNet.Diagnostics/DeveloperExceptionPage/DeveloperExceptionPageMiddleware.cs]: That;s ok

**Comment by davidfowl** [comment]
:shipit:

**Pull request closed without merging**